File: DP/PainHouseII.py
from typing import List

class Solution:
    def minCostII(self, costs: List[List[int]]) -> int:
        n = len(costs)
        if n == 0:
            return 0
        k = len(costs[0])
        if k == 0:
            return 0

        min1 = min2 = -1  # Indexes of the smallest and second smallest in the previous row

        for i in range(n):
            lastMin1, lastMin2 = min1, min2
            min1 = min2 = -1

            for j in range(k):
                if i > 0:
                    if j != lastMin1:
                        costs[i][j] += costs[i - 1][lastMin1]
                    else:
                        costs[i][j] += costs[i - 1][lastMin2]

                # Update current min1 and min2
                if min1 == -1 or costs[i][j] < costs[i][min1]:
                    min2 = min1
                    min1 = j
                elif min2 == -1 or costs[i][j] < costs[i][min2]:
                    min2 = j

        return costs[n - 1][min1]


obj = Solution()

data = """
4 2
1 2
2 1
1 2
2 1



""".strip();
lst = []
for dt in data.split("\n")[1:]:
    lst.append(list(map(int, dt.split())))
# ['3 4', '3 4 4 5', '2 4 1 4', '1 2 3 4']

print(obj.minCostII(lst))

File: DP/PainHouseII.java
package DP;

public class PainHouseII { 
    public int minCostII(int[][] costs) {
        int n = costs.length;
        if (n == 0) return 0;
        int k = costs[0].length;
        if (k == 0) return 0;

        // Track the least and second least cost from the previous row
        int min1 = -1, min2 = -1;

        for (int i = 0; i < n; i++) {
            int lastMin1 = min1, lastMin2 = min2;
            min1 = -1;
            min2 = -1;

            for (int j = 0; j < k; j++) {
                if (i > 0) {
                    // If previous color was not j, use the min1 cost
                    if (j != lastMin1) {
                        costs[i][j] += costs[i - 1][lastMin1];
                    } else {
                        // If previous color was j, use second min
                        costs[i][j] += costs[i - 1][lastMin2];
                    }
                }
                // Update min1 and min2 for current row
                if (min1 == -1 || costs[i][j] < costs[i][min1]) {
                    min2 = min1;
                    min1 = j;
                } else if (min2 == -1 || costs[i][j] < costs[i][min2]) {
                    min2 = j;
                }
            }
        }

        return costs[n - 1][min1];
    }

    // For testing with console input format
    public static void main(String[] args) {
        int[][] costs = {
            
            {8, 2},
            {4, 3},
            {1, 7}

        };

        PainHouseII solver = new PainHouseII();
        System.out.println(solver.minCostII(costs)); // Output: 5
    }
}


public static void displayUserInfo(String name, int age)

public static void performOperations(int a, int b)

public static String checkEvenOdd(int n)

public static int findLargest(int a, int b, int c)

public static void calculateGrade(int[] marks)

public Employee(int id, String name, double salary)
public int getId()
public String getName()
public double getSalary()







